docs/development/design/build-images.rst

```diff
@@ -0,0 +1,216 @@
+
+There are some dependencies that are not easy to update and keep compatibility with all the users at the same time.
+Upgrading ``nodejs`` may make lot of old projects expecting the older version to start failing all their builds.
+On the other hand, sticking with an old version avoid users requiring a newer version to build their documentation.
+To handle this case and others, we have been thinking on supporting custom Docker images.
+
+It's not clear to me how it would be the implementation of this, but I see different paths to discuss and explore:
+
+#. Allow a ``build.dockerfile`` config pointing to a ``Dockerfile``
+  * ``FROM readthedocs/build:ubuntu20`` is required to be a valid image (to share layers)
+  * the image is build each time a build is triggered consuming build time
+#. Create a branch per custom image in ``readthedocs-docker-images`` repository
+  * use ``ubuntu20`` as base image and add the custom extra requirements
+  * build the image using our current process (Docker Hub)
+  * add the custom image to our ``-ops`` repository
+  * re-build builders to pull down the new custom image
+  * set the project to use this custom image, eg. ``readthedocs/build:<project-slug>``
+
+
+Updating versions over time
+---------------------------
+
+How do we add/upgrade a Python version?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Python patch versions can be upgraded and backported to all the images without problems.
+There is only needed to rebuild ``ubuntu20`` and most of the layers will remain shared with ``-base`` and ``-pdf``.
+
+In case we need to *add* a new Python version, the situation is similar.
+We can add the new version by using ``pyenv`` and rebuilding the ``ubuntu20`` image.
+
+
+How do we upgrade system versions?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We usually don't upgrade these dependencies unless we upgrade the Ubuntu version.
+So, they will be only upgraded when we go from Ubuntu 18.04 LTS to Ubuntu 20.04 LTS for example.
+
+Examples of these versions are:
+
+* doxygen
+* git
+* subversion
+* pandoc
+* nodejs / npm
+* swig
+* rust
+
+
+How do we add an extra requirement?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If a user asks for a new requirement (eg. azure CLI, ``az`` command) it should go into the
+"user requirements" section in the ``ubuntu20-base`` image.
+However, that will force us to rebuild all the images.
+
+We could use the section named as "future user extra requirements" for this,
+and it will force us to only rebuild the ``ubuntu20`` image.
+
+Both approaches will require to rebuild all the custom docker images from our users/customers
+that are based on the ``ubuntu20`` image.
+
+
+How do we remove an old Python version?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+At some point an old version of Python will be deprecated (eg. 3.4) and will be removed from our Docker images.
+These versions should only be removed when the OS in the ``base`` is upgraded (eg. from ``ubuntu20`` to ``ubuntu22``).
```

Comment:
This seems like a good reason to have explicit python versioned images.

Reply:
How long do we want to support old Python versions? That's the question we need to answer. If we have explicit python versioned images, _when_ are going to remove the image `ubuntu18-py34` from servers?

The _when_ what we need a plan for. Here, I'm suggesting doing this _breaking change_ together with other _breaking changes_ (eg. OS upgrade) but we can keep the same Python versions from `ubuntu20` in `ubuntu22` if we want.